### PR TITLE
grouping in granular moe 

### PR DESCRIPTION
**Pull request by ludziej**

### CODE REVIEW

File: lizrd/support/logging.py [hunk 1]
```diff
@@ -108,9 +108,10 @@ def potentially_log_plotly_figure_scalars(
                 title=title, series=series_std, value=std, iteration=iteration
             )
         else:
-            print(
-                f"Could not log scalars for plotly figure of type {type(figure.data[0])}"
-            )
+            pass  # commented it out because it's too verbose and can't debug
```
Comment (kuba-krj) [inline lizrd/support/logging.py:111]: I think we can totally remove this print, no need to comment out

**Review by kuba-krj — COMMENTED** [review]
A test that verifies implementation with `one_hot` works the same as with `index_select` and `index_add` would be great

File: research/conditional/moe_layers/expert_choice.py [hunk 2]
```diff
@@ -21,6 +21,8 @@ def __init__(
         expert_size: int,
         topk_fraction: float,
         random_perm: bool = False,
+        group_by_batch: bool = False,
+        one_hot_impl: bool = False,
         softmax_over: Literal["tokens", "experts"] = "tokens",
         group_granular_moe_by_batch: bool = False,
```
Comment (kuba-krj) [inline research/conditional/moe_layers/expert_choice.py:27]: Is this option used anywhere now?

File: research/conditional/moe_layers/expert_choice.py [hunk 3]
```diff
@@ -76,9 +90,12 @@ def forward(self, x: torch.Tensor):
             )
             # transform such that first dimension corresponds to experts
             gate_out = gate_out.permute(2, 0, 1)
-            # flatten batch_size x seq_len
-            self.update_cache_for_logging("unflatten_gate_out", gate_out)
-            gate_out = gate_out.flatten(start_dim=1)
+            # flatten batch_size x seq_lenging("unflatten_gate_out", gate_out)
```
Comment (kuba-krj) [inline research/conditional/moe_layers/expert_choice.py:93]: Is this line correct? Looks like two lines are glued together

File: research/conditional/moe_layers/expert_choice.py [hunk 4]
```diff
@@ -21,8 +21,10 @@ def __init__(
         expert_size: int,
         topk_fraction: float,
         random_perm: bool = False,
+        group_by_batch: bool = False,
+        one_hot_impl: bool = False,
+        softmax_ungrouped: bool = False,
```
Comment (sebastianjaszczur) [inline research/conditional/moe_layers/expert_choice.py:26]: wolałbym "softmax_grouped", bez negacji. Bo później mamy podwójną negację przy not_softmax_ungrouped, i ciężko się czyta.

File: research/conditional/moe_layers/expert_choice.py [hunk 5]
```diff
@@ -59,26 +67,48 @@ def __init__(
         ).requires_grad_(True)
         self.ln = LayerNorm(dmodel)
         assert softmax_over in ["tokens", "experts"]
+        assert not self.softmax_ungrouped or self.group_by_batch
```
Comment (sebastianjaszczur) [inline research/conditional/moe_layers/expert_choice.py:70]: dlaczego ten if? potencjalnie można by ustawić niezależnie te rzeczy. Potencjalnie ustawiając oba na "None" domyślnie, i jeśli jeden jest ustawiony to ustawiać oba.

File: research/conditional/moe_layers/expert_choice.py [hunk 6]
```diff
@@ -59,26 +67,48 @@ def __init__(
...
             )
-            # transform such that first dimension corresponds to experts
-            gate_out = gate_out.permute(2, 0, 1)
-            # flatten batch_size x seq_len
-            self.update_cache_for_logging("unflatten_gate_out", gate_out)
-            gate_out = gate_out.flatten(start_dim=1)
+        self.update_cache_for_logging("unflatten_gate_out", gate_out)
+
+        # each expert chooses k within dimension 1
+        if not self.group_by_batch and not self.softmax_ungrouped:
```
Comment (sebastianjaszczur) [inline research/conditional/moe_layers/expert_choice.py:110]: szczególnie tutaj widzę, że grupowanie softmaxa i grupowanie wybierania mogłyby być niezależne i logika by się uprościła

File: research/conditional/train/cc_train.py [hunk 7]
```diff
@@ -71,6 +71,9 @@ def main(
     VOCAB_SIZE = 30522 if args.model_type == "bert" else 50258
     DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
 
+    if args.detect_anomaly:
+        torch.autograd.set_detect_anomaly(True)
```
Comment (sebastianjaszczur) [inline research/conditional/train/cc_train.py:75]: co to wgl jest? to wygląda na niezależną zmianę, co nie?

File: research/conditional/utils/argparse.py [hunk 8]
```diff
@@ -85,7 +86,13 @@ def introduce_parser_arguments(parser):
     )
     parser.add_argument("--data_distributed", action="store_true")
     parser.add_argument("--group_granular_moe_by_batch", action="store_true")
+    parser.add_argument("--granuar_moe_one_hot_impl", action="store_true")
```
Comment (sebastianjaszczur) [inline research/conditional/utils/argparse.py:89]: literówka: granular , not granuar

**Review by sebastianjaszczur — CHANGES_REQUESTED** [review]
Nie skończyłem review jeszcze, ale jakieś komentarze. Jak coś to pytaj na żywo.

File: research/conditional/moe_layers/expert_choice.py (hunk 4 above)
Comment (ludziej) [reply research/conditional/moe_layers/expert_choice.py:26]: to jest flaga na inputcie, one domyślnie są na false w konwencji całego naszego repo. I chyba jest to w miare czytelne, bo naturalnie wszystko jest "False", a "True" to dodatkowy feature włączony.

File: research/conditional/train/cc_train.py (hunk 7 above)
Comment (ludziej) [reply research/conditional/train/cc_train.py:75]: po częśći niezależna, ale była używana do debugowania niestabilności grupowania

File: research/conditional/moe_layers/expert_choice.py (hunk 5 above)
Comment (ludziej) [reply research/conditional/moe_layers/expert_choice.py:70]: tak spk tak lepiej

File: lizrd/support/logging.py [hunk 9]
```diff
@@ -108,9 +108,7 @@ def potentially_log_plotly_figure_scalars(
                 title=title, series=series_std, value=std, iteration=iteration
             )
         else:
-            print(
-                f"Could not log scalars for plotly figure of type {type(figure.data[0])}"
-            )
+            pass  # removed warning because it's too verbose and can't debug
```
[the remark this answers is not in the document]
Comment (ludziej) [reply lizrd/support/logging.py:111]: tzn ja nie wiem czy kiedykolwiek go potrzebujemy bo to nie jest bug, @kuba-krj czy on ma wgl sens?

**Comment by ludziej** [comment]
dopisałem jeszcze alternatywną implementacje która squashuje wszystkie nieliniowości w jednego einsuma, jak i testy jednostkowe za pomysłem @kuba-krj które porównują czy obie impementacje są równoznaczne z poprzednią

**Comment by ludziej** [comment]
https://app.neptune.ai/o/pmtest/org/llm-random/runs/compare?viewId=standard-view&dash=loss-detailed-99e799d0-c13b-46d6-83f8-d1dfbc4e44e1&compare=IwBgnANMUMwQLANiA

tutaj porównanie na bercie: granular zwykłe, grouping, grouping z one_hot i grouping z maksymalnymi einsumami

one hot jak widać przyspiesza o około 10%, maxymalne einsumy przyspieszają jeszcze bardziej, ale obie metody dodają odrobine pamięci. Maksymalne einsumy dodaja na tyle dużo, że są uruchomione z batchem (512-64), ale działaja za to dodatkowe 10% szybciej (ale z mniejszym batchem o 1/8)

Wynika to z tego, że einsum z stragety=optimal, optymalizuje złożoność kosztem również pamięci, można będzie to zmienić jeśli chcemy, tak mi sie wydaje.

**Comment by ludziej** [comment]
Efektywnie sama warstwa z full_einsum działa ponad 2x szybciej w tej implementacji.

File: lizrd/support/logging.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kuba-krj) [reply lizrd/support/logging.py:111]: Moim zdaniem ten print jest kompletnie niepotrzebny

[the remark this answers is not in the document]
Comment (ludziej) [reply lizrd/support/logging.py:111]: dobra to usuwam całkowicie

**Review by kuba-krj — APPROVED** [review]
LGTM, ale poczekałbym jeszcze na approve Sebastiana

**Review by sebastianjaszczur — APPROVED** [review]
LGTM. Chociaż nie czuję się pewny z logiką tego wszystkiego.
